### PR TITLE
TEPHRA-53 Add HBase 1.0 compatibility

### PR DESCRIPTION
**Pull request by ghelmling**
This adds two new compatibility modules to support the two new versions of HBase:
* tephra-hbase-compat-1.0 - supports Apache HBase 1.0.x releases
* tephra-hbase-compat-1.0-cdh - supports the modified version of HBase shipped in CDH 5.4

Since the version of HBase shipped in CDH 5.4 is neither binary nor source compatible with Apache HBase 1.0, it needs to be supported as a separate build artifact.

### CODE REVIEW

File: tephra-core/src/test/java/co/cask/tephra/util/HBaseVersionTest.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    ver = HBaseVersion.VersionNumber.create("3.2.1-baz1.2.3");
+    assertVersionNumber(ver, 3, 2, 1, "baz1.2.3", false);
+
+    ver = HBaseVersion.VersionNumber.create("3.2.1-baz1.2.3-SNAPSHOT");
+    assertVersionNumber(ver, 3, 2, 1, "baz1.2.3", true);
+
+    try {
+      ver = HBaseVersion.VersionNumber.create("abc");
+      fail("Invalid verison number 'abc' should have thrown a ParseException");
+    } catch (ParseException pe) {
```
Comment (henry-cask) [inline tephra-core/src/test/java/co/cask/tephra/util/HBaseVersionTest.java]: Should we not catch this with try-catch then?

**Comment by henry-cask** [comment]
Could we have comment indicating where the needs to have CDH specific code as Javadoc? I am seeing the CDH compatibility module but find it hard to find in the code diff which methods or code needed to be modified from support for Apache HBase 1.0

File: tephra-core/src/test/java/co/cask/tephra/util/HBaseVersionTest.java (hunk 1 above)
Comment (ghelmling) [inline tephra-core/src/test/java/co/cask/tephra/util/HBaseVersionTest.java]: Not sure what you mean -- that we shouldn't throw the ParseException in VersionNumber.create(), but should return something else instead; or that we should move this check to a separate test with an annotation of @Test(expected=ParseException)?

**Comment by ghelmling** [comment]
There are no modified methods in our implementations of the two modules.

What is different between Apache HBase 1.0 and CDH HBase 1.0 is the return types of various methods in the HBase Scan, Get, Delete, HTableDescriptor, HColumnDescriptor classes.  This means that if we only had 1 hbase-compat-1.0 module, it would only run on binaries from one of Apache HBase 1.0 or CDH 5.4 (whichever it was compiled against), and fail to run on binaries from the other.

We could skirt around this by using reflection to call any methods on these classes.  But they are fairly core classes, and from past experience using reflection in similar scenarios, it becomes difficult to maintain this code.

The API compatibility report showing the changes is an attachment on TEPHRA-53: https://issues.cask.co/secure/attachment/11016/hbase_1.0.0_vs_cdh5.4.2-release_compat_report.html

File: tephra-core/src/test/java/co/cask/tephra/util/HBaseVersionTest.java (hunk 1 above)
Comment (henry-cask) [inline tephra-core/src/test/java/co/cask/tephra/util/HBaseVersionTest.java]: Ah yes, so sorry about the confusion of my comment. I meant to annotate the test with  @Test(expected=ParseException) as separate method. 

Not sure if it is too much to separate as separate method though.

Comment (ghelmling) [inline tephra-core/src/test/java/co/cask/tephra/util/HBaseVersionTest.java]: Ah, I see.  It didn't really seem worth it to separate this and the one below into their own tests.  Logically, these all seem like part of the same overall test to me.

**Comment by henry-cask** [comment]
Thanks for the links, I understand what you mean.

**Comment by henry-cask** [comment]
I did a pass and did not see anything glaringly wrong. 
Mostly to accommodate CDH version of HBase API changes.

LGTM

**Pull request merged**